### PR TITLE
os: bcast address interface

### PR DESCRIPTION
**Pull request by Txiaozhe**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by Txiaozhe** [comment]
#414

File: deps/libtuv/src/unix/darwin.c [hunk 1]
```diff
@@ -293,6 +293,14 @@ int uv_interface_addresses(uv_interface_address_t** addresses, int* count) {
       address->netmask.netmask4 = *((struct sockaddr_in*) ent->ifa_netmask);
     }
 
+    if (ent->ifa_dstaddr) {
+      if (ent->ifa_dstaddr->sa_family == AF_INET6) {
+        address->broadcast.broadcast6 = *((struct sockaddr_in6*) ent->ifa_dstaddr);
+      } else {
+        address->broadcast.broadcast4 = *((struct sockaddr_in*) ent->ifa_dstaddr);
+      }
+    }
```
Comment (yorkie) [inline deps/libtuv/src/unix/darwin.c]: You should change the other platforms, too.

File: test/run_pass/test_os_network_interfaces.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+var os = require('os');
+
+console.log(os.networkInterfaces());
```
Comment (yorkie) [inline test/run_pass/test_os_network_interfaces.js]: You should assert the fields, just like what the Node.js does at https://github.com/nodejs/node/blob/master/test/parallel/test-os.js#L131-L172.

**Comment by yorkie** [comment]
Label this as `push upstream`, so when this gets done, I suggest that @Txiaozhe you could push this change on libuv and Node.js core :)

File: test/run_pass/test_os_network_interfaces.js (hunk 2 above)
Comment (Txiaozhe) [reply test/run_pass/test_os_network_interfaces.js]: Yes, I will

File: test/run_pass/test_os_network_interfaces.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+var assert = require('assert');
+var os = require('os');
+
+var family = {
+  IPv4: 1,
+  IPv6: 1
+};
+
+var net_info = os.networkInterfaces();
```
Comment (yorkie) [inline test/run_pass/test_os_network_interfaces.js]: Ensure the interfaces array's length is >0 :)

File: test/run_pass/test_os_network_interfaces.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+for (var net_name in net_info) {
+  net_info[net_name].forEach(net => {
```
Comment (yorkie) [inline test/run_pass/test_os_network_interfaces.js]: net_name -> name

**Comment by yorkie** [comment]
By the way, you should add your test case to https://github.com/Rokid/ShadowNode/blob/master/test/testsets.json in alphabeta.

**Comment by Txiaozhe** [comment]
Thanks, has been fixed.

File: test/run_pass/test_os_network_interfaces.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+for (var name in net_info) {
+  assert(net_info[name].length);
+  net_info[name].forEach(net => {
+    assert.equal(typeof net.address, 'string');
+    assert.equal(typeof net.netmask, 'string');
+    assert.equal(family[net.family], 1);
+    assert.equal(typeof net.broadcast, 'string');
+    assert.equal(typeof net.mac, 'string');
+  });
+}
```
Comment (yorkie) [inline test/run_pass/test_os_network_interfaces.js]: Could you please just move the Node.js [test](https://github.com/nodejs/node/blob/master/test/parallel/test-os.js#L131-L172) into this repository?

Comment (Txiaozhe) [reply test/run_pass/test_os_network_interfaces.js]: Yes, I will try it.

File: test/common/index.js [hunk 6]
```diff
@@ -61,6 +89,18 @@ function expectsError(fn, exact) {
...
+  isLinuxPPCBE: isLinuxPPCBE,
+  isSunOS: isSunOS,
+  isFreeBSD: isFreeBSD,
+  isOpenBSD: isOpenBSD,
+  isLinux: isLinux,
+  isOSX: isOSX,
+  isOSXMojave: isOSXMojave,
+
+  is: is,
+
```
Comment (yorkie) [inline test/common/index.js]: @Txiaozhe No need to add these lines.

Comment (Txiaozhe) [reply test/common/index.js]: Awkward! I will do it.

Comment (yorkie) [reply test/common/index.js]: Never mind, you are always welcome :)

File: test/run_pass/test_os.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+var platform = os.platform();
+common.is.string(platform);
+assert.ok(platform.length > 0);
+
+var EOL = os.EOL;
+if (common.isWindows) {
+  assert.strictEqual(EOL, '\r\n');
+} else {
+  assert.strictEqual(EOL, '\n');
+}
```
Comment (yorkie) [inline test/run_pass/test_os.js]: We do not need the above tests at this PR, could you please remove them?

File: test/run_pass/test_os.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    
+    var expected = [{
+      address: '127.0.0.1',
+      netmask: '255.0.0.0',
+      family: 'IPv4',
+      broadcast: "127.0.0.1",
+      mac: '00:00:00:00:00:00'
+    }];
+
+    assert.deepStrictEqual(actual, expected);
```
Comment (yorkie) [inline test/run_pass/test_os.js]: Unix/Linux also needs to cover :)

File: src/modules.json [hunk 9]
```diff
@@ -56,7 +56,11 @@
     },
     "assert": {
       "js_file": "js/assert.js",
-      "require": ["util"]
+      "require": ["util", "comparisons"]
+    },
+    "comparisons": {
```
Comment (legendecas) [inline src/modules.json]: An internal module doesn't been expected to be in public scope. An internal prefix might help.

Comment (Txiaozhe) [reply src/modules.json]: Yes, I change it to internal_comparisons. Thanks! :)

File: test/run_pass/test_os.js (hunk 7 above)
Comment (Txiaozhe) [reply test/run_pass/test_os.js]: I will do!

File: test/run_pass/test_os.js (hunk 8 above)
Comment (Txiaozhe) [reply test/run_pass/test_os.js]: You mean I should comparisons Unix/Linux such as:
```c
if (Unix) {
    ....
} else if (Linux) {
    ....
}
```
?

Comment (yorkie) [reply test/run_pass/test_os.js]: Nope, you are only testing this function on MacOS.

Comment (Txiaozhe) [reply test/run_pass/test_os.js]: Awkward, I found my fault~

File: src/internal/util/comparisons.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  if (val1 && val1 === val2) {
+    return true;
+  }
+
+  return strictDeepEqual(val1, val2);
+}
+
+module.exports = {
+  isDeepStrictEqual: isDeepStrictEqual
+};
```
Comment (yorkie) [inline src/internal/util/comparisons.js]: Also, no need to add this file, please revert it, and make every pull request do simply.

**Comment by Txiaozhe** [comment]
> Label this as `push upstream`, so when this gets done, I suggest that @Txiaozhe you could push this change on libuv and Node.js core :)

This feature was discussed in [node issues: 23437](https://github.com/nodejs/node/issues/23437), but it seems to have been rejected.

**Comment by yorkie** [comment]
@Txiaozhe Nope, it's still open for changes, I think that we could continue to work on @pprindeville's patch with the comment from @addaleax :)

**Comment by Txiaozhe** [comment]
Yes, I will try!

**Comment by yorkie** [comment]
Also, please clean up your branch firstly, we won't merge patch which commits > 10 :)

**Comment by Txiaozhe** [comment]
> Also, please clean up your branch firstly, we won't merge patch which commits > 10 :)

Yes, I will.

**Pull request closed without merging**